torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1147,7 +1147,35 @@ void InitXlaModuleBindings(py::module m) {
   m.def("_xla_get_tensor_id",
         [](const at::Tensor& tensor) { return GetTensorId(tensor); });
   m.def("_xla_get_spmd_config_is_locked", []() { return GetLockSpmdConfig(); });
+  m.def("_xla_force_spmd_device", []() {
+    // It is actually more easier to force SPMD mode than blocking if there is
+    // non-SPMD initialized tensors, for the 3rd-party solution integration. For
+    // instance, HuggingFace trainer pre-loads embeddings table and the training
+    // initialization is done over multiple scripts. Being able to force SPMD
+    // allows the users to call `xr.use_spmd()` more freely, given that the
+    // earlier they call, the smaller the one-time overhead of replicating
+    // non-SPMD backed tensors.
+    torch::lazy::BackendDevice backend_device = bridge::GetCurrentDevice();
+    std::vector<XLATensorPtr> xtensors =
+        XLAGraphExecutor::Get()->GetLiveTensors(&backend_device);
+    for (auto xtensor : xtensors) {
+      XlaDeviceType xla_device_type =
+          static_cast<XlaDeviceType>(xtensor->GetDevice().type());
+      if (xla_device_type != XlaDeviceType::SPMD) {
+        // TODO(yeounoh) this can be further optimized via CopyToDevice.
+        at::Tensor tensor = xtensor->ToTensor(false);
+        xtensor->SetXlaData(
+            TensorToXlaData(tensor, ParseDeviceString("SPMD:0")));
+      }
+    }
+  });
   m.def("_init_computation_client", []() { runtime::GetComputationClient(); });
+  m.def("_xla_get_device_hw_type", [](const at::Tensor& tensor) {
```

Comment:
I think we have something like this already: https://github.com/pytorch/xla/blob/b6b9c6dabe6359196596a9890e60d6c15c0b7a7d/torch_xla/core/xla_model.py#L237

Can we do some consolidation?

Reply:
Oh, thank you -- will look into this!

Reply:
@alanwaketan it turns out that `xla_device_hw` tracked `PJRT_DEVICE` value instead. I think there is room for improvement/consolidation. Will follow-up in a separate PR.

Reply:
Sure.